### ISSUE TITLE
`model_parameters()` fails for survreg

### ISSUE THREAD

**Issue by strengejacke** [easystats/parameters#673]
Needed to be fixed for requested CRAN submission, somewhat urgent (breaks _report_).

``` r
library(survival)
library(parameters)
set.seed(123)
mod_survreg <-
  survival::survreg(
    formula = Surv(futime, fustat) ~ ecog.ps + rx,
    data = ovarian,
    dist = "logistic"
  )

model_parameters(mod_survreg)
#> Error: Sorry, `model_parameters()` failed with the following error (possible class
#>   'survreg' not supported):
#> 
#>   evaluation nested too deeply: infinite recursion / options(expressions=)?
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
@vincentarelbundock I think this is due to your changes to the `vcov` argument. In https://github.com/easystats/parameters/blob/main/R/methods_survival.R, `p_value.survreg()` creates an infinite loop:

```
p_value.survreg <- function(model, method = NULL, ...) {
  if (isTRUE(list(...)$robust) || "vcov" %in% names(list(...))) {
    return(p_value(model, ...))
  }
  s <- summary(model)
  p <- s$table[, "p"]
  .data_frame(
    Parameter = .remove_backticks_from_string(names(p)),
    p = as.vector(p)
  )
}
```

Can you look into this?

**Comment by strengejacke** [comment]
Same might happen to `p_value.zeroinfl()` in https://github.com/easystats/parameters/blob/main/R/methods_pscl.R

**Comment by strengejacke** [comment]
pscl seems to work:

``` r
library(pscl)
#> Classes and Methods for R developed in the
#> Political Science Computational Laboratory
#> Department of Political Science
#> Stanford University
#> Simon Jackman
#> hurdle and zeroinfl functions by Achim Zeileis
library(parameters)
data("bioChemists")
m1 <- zeroinfl(art ~ fem + mar + kid5 + ment | kid5 + phd, data = bioChemists)

p_value(m1, method = "robust")
#>           Parameter            p     Component
#> 1 count_(Intercept) 1.421165e-16   conditional
#> 2    count_femWomen 9.234769e-05   conditional
#> 3  count_marMarried 3.832740e-02   conditional
#> 4        count_kid5 6.001337e-04   conditional
#> 5        count_ment 8.829659e-24   conditional
#> 6  zero_(Intercept) 3.211149e-02 zero_inflated
#> 7         zero_kid5 8.306852e-01 zero_inflated
#> 8          zero_phd 6.539337e-02 zero_inflated
p_value(m1, method = "robust", vcov = NULL)
#>           Parameter            p     Component
#> 1 count_(Intercept) 1.421165e-16   conditional
#> 2    count_femWomen 9.234769e-05   conditional
#> 3  count_marMarried 3.832740e-02   conditional
#> 4        count_kid5 6.001337e-04   conditional
#> 5        count_ment 8.829659e-24   conditional
#> 6  zero_(Intercept) 3.211149e-02 zero_inflated
#> 7         zero_kid5 8.306852e-01 zero_inflated
#> 8          zero_phd 6.539337e-02 zero_inflated
p_value(m1, vcov = NULL)
#>           Parameter            p     Component
#> 1 count_(Intercept) 1.421165e-16   conditional
#> 2    count_femWomen 9.234769e-05   conditional
#> 3  count_marMarried 3.832740e-02   conditional
#> 4        count_kid5 6.001337e-04   conditional
#> 5        count_ment 8.829659e-24   conditional
#> 6  zero_(Intercept) 3.211149e-02 zero_inflated
#> 7         zero_kid5 8.306852e-01 zero_inflated
#> 8          zero_phd 6.539337e-02 zero_inflated
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
@vincentarelbundock robust estimation doesn't seem to work for `zeroinfl` models (not supported). Couldn't we just remove the "robust" argument?

**Comment by strengejacke** [comment]
I think the last commit should fix this, however, it seems there is an issue with the sandwich package, which _should_ support these models (but does not?)

``` r
library(pscl)
#> Classes and Methods for R developed in the
#> Political Science Computational Laboratory
#> Department of Political Science
#> Stanford University
#> Simon Jackman
#> hurdle and zeroinfl functions by Achim Zeileis
library(parameters)
data("bioChemists")
m1 <- zeroinfl(art ~ fem + mar + kid5 + ment | kid5 + phd, data = bioChemists)

p_value(m1)
#>           Parameter            p     Component
#> 1 count_(Intercept) 1.421165e-16   conditional
#> 2    count_femWomen 9.234769e-05   conditional
#> 3  count_marMarried 3.832740e-02   conditional
#> 4        count_kid5 6.001337e-04   conditional
#> 5        count_ment 8.829659e-24   conditional
#> 6  zero_(Intercept) 3.211149e-02 zero_inflated
#> 7         zero_kid5 8.306852e-01 zero_inflated
#> 8          zero_phd 6.539337e-02 zero_inflated
p_value(m1, robust = TRUE)
#> Warning: The `robust` argument is superseded by the `vcov` argument.
#> Error in ef/X: non-conformable arrays

standard_error(m1)
#>           Parameter          SE     Component
#> 1 count_(Intercept) 0.067973929   conditional
#> 2    count_femWomen 0.058683970   conditional
#> 3  count_marMarried 0.065928261   conditional
#> 4        count_kid5 0.048737645   conditional
#> 5        count_ment 0.002120739   conditional
#> 6  zero_(Intercept) 0.433782768 zero_inflated
#> 7         zero_kid5 0.215095486 zero_inflated
#> 8          zero_phd 0.135200903 zero_inflated
standard_error(m1, robust = TRUE)
#> Warning: The `robust` argument is superseded by the `vcov` argument.
#> Error in ef/X: non-conformable arrays

sandwich::vcovHC(m1)
#> Error in ef/X: non-conformable arrays


library(survival)
library(parameters)
set.seed(123)
m1 <- survival::survreg(
  formula = Surv(futime, fustat) ~ ecog.ps + rx,
  data = ovarian,
  dist = "logistic"
)

p_value(m1)
#>     Parameter             p
#> 1 (Intercept)  2.270987e-01
#> 2     ecog.ps  4.234150e-01
#> 3          rx  2.224287e-01
#> 4  Log(scale) 6.506721e-129
p_value(m1, robust = TRUE)
#> Warning: The `robust` argument is superseded by the `vcov` argument.
#> Error in ef/X: non-conformable arrays

standard_error(m1)
#>     Parameter          SE
#> 1 (Intercept) 552.5705685
#> 2     ecog.ps 263.0648320
#> 3          rx 262.3595438
#> 4  Log(scale)   0.2408162
standard_error(m1, robust = TRUE)
#> Warning: The `robust` argument is superseded by the `vcov` argument.
#> Error in ef/X: non-conformable arrays

sandwich::vcovHC(m1)
#> Error in ef/X: non-conformable arrays

model_parameters(m1)
#> # Fixed Effects
#> 
#> Parameter   | Log-Odds |     SE |             95% CI |     z |      p
#> ---------------------------------------------------------------------
#> (Intercept) |   667.43 | 552.57 | [-415.59, 1750.45] |  1.21 | 0.227 
#> ecog ps     |  -210.59 | 263.06 | [-726.18,  305.01] | -0.80 | 0.423 
#> rx          |   320.10 | 262.36 | [-194.11,  834.32] |  1.22 | 0.222 
#> Log(scale)  |     5.82 |   0.24 | [   5.35,    6.29] | 24.16 | < .001
#> 
#> Uncertainty intervals (equal-tailed) and p values (two-tailed) computed using a
#>   Wald z-distribution approximation.
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
@vincentarelbundock I think I fixed the main issue here. However, models from package *pscl* do currently not work with robust methods. *sandwich* seems to fail for those models (see above, despite the docs saying that it should work).

Futhermore, models from *pscl*  have a `Component` column (conditional and zero_inflated), while `standard_error.default()` and `p_value.default()`  did not return that column. Previously, when `vcov` is given, those methods did not include the component column. I changed the code so that these methods now include that column if necessary. This is only required for robust-estimation, because all model-specific methods will call the default method for robust estimation. I hope this works now (can't check, because sandwich fails)

**Comment by strengejacke** [comment]
ok, got it. not all functions are working, so I test for specific vcov-funs now.

**Comment by strengejacke** [comment]
I think this is resolved now, but we should still add more tests here:
https://github.com/easystats/parameters/blob/main/tests/testthat/test-robust.R

**Comment by vincentarelbundock** [comment]
All of this happened while I was sleeping?!

Sounds like things are resolved, but please let me know if there's anything I should look at. (Note that I probably won't have time to do much in the next couple days.)

**Comment by strengejacke** [comment]
I think everything is resolved now, adding some tests...

- For some methods (like `p_value.survreg()`), we had an infinite loop when using `vcov`, because it call `p_value()` from `p_value()` (instead of `p_value.default()`)
- When models from _pscl_ were passed to `standard_error.default()` or `p_value.default()`, no `"Component"` column was returned, so `model_parameters()` did not work for those models when `vcov` was used.
- I found some minor copy/paste errors (`p_value.polr()` called `standard_error()` when `vcov` was used).
- `ci.polr()` ignored `vcov`.

All of these should have been fixed. I'm not adding tests for betareg and ivreg. It would be good, in general (and not immediately), to have tests for models supported by *(club)sandwich* in the `test-robust.R` file.
